storage/integration_test.go

```diff
@@ -3888,6 +3888,261 @@ func TestIntegration_LockBucket_MetagenerationRequired(t *testing.T) {
+	multiTransportTest(ctx, t, func(t *testing.T, ctx context.Context, _ string, prefix string, client *Client) {
+		setTrue, setFalse := true, false
+
+		for _, test := range []struct {
+			desc                string
+			objectRetentionMode string
+			enable              *bool
+			wantRetentionMode   string
+		}{
+			{
+				desc:              "ObjectRetentionMode is not enabled by default",
+				wantRetentionMode: "",
+			},
+			{
+				desc:                "Setting just ObjectRetentionMode should not work",
+				objectRetentionMode: "Enabled",
+				wantRetentionMode:   "",
+			},
+			{
+				desc:                "ObjectRetentionMode is ignored",
+				objectRetentionMode: "disabled",
+				enable:              &setTrue,
+				wantRetentionMode:   "Enabled",
+			},
+			{
+				desc:              "Empty ObjectRetentionMode",
+				enable:            &setTrue,
+				wantRetentionMode: "Enabled",
+			},
+			{
+				desc:              "Empty ObjectRetentionMode set to false",
+				enable:            &setFalse,
+				wantRetentionMode: "",
+			},
+		} {
+			t.Run(test.desc, func(t *testing.T) {
+				b := client.Bucket(prefix + uidSpace.New())
+				if test.enable != nil {
+					b = b.SetObjectRetention(*test.enable)
+				}
+
+				attrs := &BucketAttrs{
+					ObjectRetentionMode: test.objectRetentionMode,
+				}
+
+				err := b.Create(ctx, testutil.ProjID(), attrs)
+				if err != nil {
+					t.Fatalf("error creating bucket: %v", err)
+				}
+				t.Cleanup(func() { b.Delete(ctx) })
+
+				attrs, err = b.Attrs(ctx)
+				if err != nil {
+					t.Fatalf("b.Attrs: %v", err)
+				}
+				if got, want := attrs.ObjectRetentionMode, test.wantRetentionMode; got != want {
+					t.Errorf("expected ObjectRetentionMode to be %q, got %q", want, got)
+				}
+			})
+		}
+	})
+}
+
+func TestIntegration_ObjectRetention(t *testing.T) {
+	ctx := skipJSONReads(skipGRPC("not yet available in gRPC - b/308194853"), "no reads in test")
+	multiTransportTest(ctx, t, func(t *testing.T, ctx context.Context, _ string, prefix string, client *Client) {
+		h := testHelper{t}
+
+		b := client.Bucket(prefix + uidSpace.New()).SetObjectRetention(true)
+
+		if err := b.Create(ctx, testutil.ProjID(), nil); err != nil {
+			t.Fatalf("error creating bucket: %v", err)
+		}
+		t.Cleanup(func() { b.Delete(ctx) })
+
+		// Create an object with future retain until time
+		retentionUnlocked := &ObjectRetention{
+			Mode:        "Unlocked",
+			RetainUntil: time.Now().Add(time.Minute * 20).Truncate(time.Second),
+		}
+		retentionLocked := &ObjectRetention{
+			Mode:        "Locked",
+			RetainUntil: time.Now().Add(time.Minute * 20).Truncate(time.Second),
+		}
+		retentionUnlockedExtended := &ObjectRetention{
+			Mode:        "Unlocked",
+			RetainUntil: time.Now().Add(time.Hour).Truncate(time.Second),
+		}
+		retentionLockedExtended := &ObjectRetention{
+			Mode:        "Locked",
+			RetainUntil: time.Now().Add(time.Hour).Truncate(time.Second),
+		}
```

Comment:
Do all these retention times pose an issue for being able to clean up the buckets created by this test?

Reply:
I discussed this with Jesse, we don't need to explicitly test Locked retentions since they cause these cleanup issues. I've removed them and everything should clean up appropriately now.

Reply:
Excellent, thanks for confirming.